relayer/src/worker/connection.rs

```diff
@@ -0,0 +1,127 @@
+use std::{thread, time::Duration};
+
+use anomaly::BoxError;
+use crossbeam_channel::Receiver;
+use tracing::{debug, warn};
+
+use crate::connection::Connection as RelayConnection;
+use crate::telemetry::Telemetry;
+use crate::{
+    chain::handle::ChainHandlePair, object::Connection, util::retry::retry_with_index,
+    worker::retry_strategy,
+};
+
+use super::WorkerCmd;
+
+pub struct ConnectionWorker {
+    connection: Connection,
+    chains: ChainHandlePair,
+    cmd_rx: Receiver<WorkerCmd>,
+
+    #[allow(dead_code)]
+    telemetry: Telemetry,
+}
+
+impl ConnectionWorker {
+    pub fn new(
+        connection: Connection,
+        chains: ChainHandlePair,
+        cmd_rx: Receiver<WorkerCmd>,
+        telemetry: Telemetry,
+    ) -> Self {
+        Self {
+            connection,
+            chains,
+            cmd_rx,
+            telemetry,
+        }
+    }
+
+    /// Run the event loop for events associated with a [`Connection`].
+    pub(crate) fn run(self) -> Result<(), BoxError> {
+        let a_chain = self.chains.a.clone();
+        let b_chain = self.chains.b.clone();
+
+        // Flag that indicates if the worker should actively resume handshake.
+        // Set on start or when event based handshake fails.
+        let mut resume_handshake = true;
+
+        loop {
```

Comment:
What happens after the full connection handshake has been established? Is the loop supposed to continue forever?

Reply:
I always find complicated control flow inside loops difficult to follow. I understand you are just following the convention before. It would be great if we can use functions to structure the control flows instead. That would make it much easier to understand the control flow. For example something like follows:

```rust
    pub(crate) fn run(self) -> Result<(), BoxError> {
        enum Action {
            Continue,
            End,
        }

        fn do_run(worker: &ConnectionWorker) -> Result<Action, BoxError> { ... }

        loop {
            match do_run(&self) {
                Ok(Action::End) => return Ok(()),
                Err(e) => {
                    error!("error: {}. retrying");
                    // or return Err if unrecoverable
                }
                _ => {}
            }
        }
  }
```

Reply:
> What happens after the full connection handshake has been established? Is the loop supposed to continue forever?

no, see #1115

Reply:
> I always find complicated control flow inside loops difficult to follow. I understand you are just following the convention before. It would be great if we can use functions to structure the control flows instead. 
...

Please open an issue for this and we will address in a separate PR.